src/plasmapy/simulation/particle_tracker/particle_tracker.py

```diff
@@ -584,6 +596,147 @@ def inner_Bethe_stopping(v, n_e):
+    def add_scattering(
+        self,
+        target: ParticleLike,
+        differential_cross_section: (
+            Callable[
+                [
+                    u.Quantity[u.rad],  # Theta
+                    u.Quantity[u.m / u.s],  # Speed
+                    ParticleLike,  # Beam
+                    ParticleLike,  # Target
+                    u.Quantity[u.K],  # Electron Temperature
+                    u.Quantity[1 / u.m**3],  # Electron Number Density
+                ],
+                u.Quantity[u.m**2 / u.sr],  # Differential cross-section
+            ]
+            | None
+        ) = None,
+        mean_square_scatter_rate: (
+            Callable[
+                [
+                    u.Quantity[u.m / u.s],  # Speed
+                    ParticleLike,  # Beam
+                    ParticleLike,  # Target
+                    u.Quantity[u.K],  # Electron Temperature
+                    u.Quantity[1 / u.m**3],  # Electron Number Density
+                ],
+                u.Quantity[
+                    u.rad**2 / u.s
+                ],  # Mean-square scatter rate in radians^2 per second
+            ]
+            | None
+        ) = None,
+    ):
+        r"""
+        Enable particle scattering by numeric integration of the provided
+        differential cross-section.
+        """
+
+        # For scattering calculations, a multidimensional interpolator must be
+        # constructed to obtain the mean-square scattering rate as a function of
+        # velocity and number density.
+
+        if not self._is_quantity_defined_on_one_grid("n_e"):
+            warnings.warn(
+                "The electron number density is not defined on any of the provided grids! "
+                "Particle scattering will not be calculated.",
+                RuntimeWarning,
+            )
+
+            return
+        elif not self._is_quantity_defined_on_one_grid("T_e"):
+            warnings.warn(
+                "The electron temperature is not defined on any of the provided grids! "
+                "Particle scattering will not be calculated.",
+                RuntimeWarning,
+            )
+
+            # Don't set `_do_scattering`. The push loop does not have to do scattering
+            # calculations
+            return
+
+        for grid in self.grids:
+            grid.require_quantities(["n_e", "T_e"], replace_with_zeros=True)
+            self._required_quantities.update({"n_e", "T_e"})
+
+        if differential_cross_section is not None:
+            self._scattering_method = "differential cross section"
+            self._scattering_routine = differential_cross_section
+        elif mean_square_scatter_rate is not None:
+            self._scattering_method = "mean square scatter rate"
+            self._scattering_routine = mean_square_scatter_rate
+
+        self._target = target
+        self._do_scattering = True
+
+        # For now the integrals will be evaluated every time step
```

Comment:
Is there an alternative to this, since the local density and temperature at each particle is different on each timestep?

Reply:
Somewhere in the code I had a block commented out where I was essentially trying to construct a multidimensional interpolator that takes in v, T_s, and n_s and returns the mean-square rate of scattering. To do this I had to come up ranges for the parameters such that every possible combination of v, T_s, and n_s would be within the "domain" of the interpolator. The ranges (sampled in logspace) are v: 0 -> c , T_s: 0 -> max(T_s), and n_s: 0 -> max(n_s) where we cheat by taking the max value of temperature and number density encountered in the relevant grid. Does that sound like a reasonable approach?

Reply:
I suppose the question is one of computation time, and consequently on how fine of a resolution for that grid is necessary for the initial computation, before the interpolation? If constructing the interpolator with sufficient resolution is fast enough, then this could work well.